tests/Feature/PoolTest.php

```diff
@@ -89,6 +90,29 @@
     expect($count)->toEqual(5);
 });
 
+test('if a pool has a failed response that should be treated as a successful response', function () {
+    $count = 0;
+
+    $pool = (new TestConnector)->pool([
+        new ErrorRequestThatShouldBeTreatedAsSuccessful,
+    ]);
+
+    $pool->withResponseHandler(function (Response $response) use (&$count) {
+        expect($response)->toBeInstanceOf(Response::class);
+        expect($response->status())->toBe(404);
+
+        $count++;
+    });
+
+    $promise = $pool->send();
+
+    expect($promise)->toBeInstanceOf(PromiseInterface::class);
+
+    $promise->wait();
+
+    expect($count)->toEqual(1);
+});
```

Comment:
Please could we also add a similar test for a synchronous request?

Reply:
the `shouldThrowRequestException` implementation has a similar issue to the `hasRequestFailed` method.

If `shouldThrowRequestException` on the request class returns `false` then an exception will still be thrown because the `shouldThrowRequestException` on the connector class will always return `true` by default if the request fails
```php
// in Response class

public function shouldThrowRequestException(): bool
{
    $pendingRequest = $this->getPendingRequest();

	// If you customize the shouldThrowRequestException method on the request
	// then that will not have a any effect unless you also override the method on
	// the connector class
    return $pendingRequest->getRequest()->shouldThrowRequestException($this)
		|| $pendingRequest->getConnector()->shouldThrowRequestException($this);
}
```

Reply:
Hmm I've had a look, and it looks like the `shouldThrowRequestException` only returns `bool` rather than the `hasRequestFailed` was `?bool` - so I don't think this needs changing.

I wonder if this is related to the send and retry functionality because in v3 every request runs through the sendAndRetry logic but only if the `maxAttempts` > 1 than this logic will be executed.

Reply:
The maxAttempts > 1 is the problem I think.